Add Memory Overrides to Streams

NewellClark opened the pull request:
In the process of implementing [Override Stream ReadAsync/WriteAsync](https://github.com/dotnet/runtime/issues/33789), I ran the analyzer against dotnet/runtime and found 7 violations. This is one of them.
I've added the required memory-based overrides, and refactored it so the array-based overrides are implemented in terms of the memory-based overrides. In order to do this, I had to add an overload of `Inflater.SetBuffer` that accepts memory. Same thing here; I refactored the array-based `SetInput` to delegate to the new memory-based overload. 

I've also fixed `RequestStream`, `ChunkedMemoryStream`, and `NetworkStreamWrapper`, which were all trivial fixes.

Should I do a separate PR for every violation I fix, or fix them all in one PR? Some of the fixes are non-trivial, and would require adding overloads to multiple dependencies used by the violating stream, which is why I figured a separate PR would be in order for each one. 

Edit: to clarify, this is the PR I'm using for all the streams that are trivial to update. I'll be doing separate PRs for the streams that are not trivial to update.

NewellClark committed:
System.IO.Compression.DeflateStream.CopyToStream

-Implemented memory-based WriteAsync in DeflateStream.CopyToStream class.

This required implementing a memory-based overload of System.IO.Inflater.SetInput(). Previously, Inflater used a GCHandle to pin the array that was passed into SetInput. I converted it to use a MemoryHandle, and changed the array-based overload of SetInput to delegate to the new Memory-based overload.

stephentoub commented on src/libraries/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs:
```diff
@@ -923,9 +923,22 @@ public override async Task WriteAsync(byte[] buffer, int offset, int count, Canc
                     throw new InvalidDataException(SR.GenericInvalidData);
```
Technically there's a breaking change here, in that we previously were returning any exception from here or from EnsureNotDisposed out through the returned task.  But in practice, in this case I don't think it really matters.  The only caller of this is going to be the source stream's CopyToAsync, and it seems really unlikely that it'll a) call this overload (the base implementation calls the memory overload), b) misuse the method (e.g. call it when disposed), AND c) do something other than directly await the call's result, in which case whether an exception is propagated directly or via the returned task isn't perceivable.  On top of that, there's a good argument to be made that at least the EnsureNotDisposed shouldn't have been wrapped in the first place.  But if we wanted to maintain compat for this super rare case, we could wrap these with Task.FromException.

NewellClark replied:
Just to clarify, do you want me to leave it as-is?

stephentoub replied:
Let's leave the EnsureNotDisposed as is, but change this line to be:
```C#
return Task.FromException(new InvalidDataException(SR.GenericInvalidData));
```

stephentoub commented on src/libraries/System.IO.Compression/src/System/IO/Compression/DeflateZLib/Inflater.cs:
```diff
@@ -182,18 +183,42 @@ public void SetInput(byte[] inputBuffer, int startIndex, int count)
             }
         }
 
+        private struct InputBufferHandle
```
I don't see why this new struct is needed (was it just to give these operations the same name as GCHandle to avoid additional changes elsewhere?).  _inputBufferHandle should just be typed as MemoryHandle.

NewellClark replied:
I made the struct to avoid having to mark more methods as `unsafe`. I can remove it if that's not a problem.

stephentoub replied:
It's not a problem to surround the regions that use pointers with unsafe (the whole method needn't be marked if that would significantly increase the scope).  Whether unsafe is there or not, a pointer is still being used; there's actually a benefit in "unsafe" showing up where the pointer is being used, as it makes it clear that something potentially dangerous is happening.

NewellClark replied:
Fixed.

stephentoub commented on src/libraries/System.IO.Compression/src/System/IO/Compression/DeflateZLib/Inflater.cs:
```diff
@@ -182,18 +183,42 @@ public void SetInput(byte[] inputBuffer, int startIndex, int count)
+            Debug.Assert(NeedsInput(), "We have something left in previous input!");
+            Debug.Assert(!_inputBufferHandle.IsAllocated);
+
+            if (inputBuffer.Length == 0)
```
```suggestion
            if (inputBuffer.IsEmpty)
```

NewellClark replied:
Fixed.

stephentoub reviewed (commented):
Thanks for working on this.

stephentoub commented:
> Should I do a separate PR for every violation I fix, or fix them all in one PR?

In particular for any that are controversial or non-trivial, one per PR is good.

For the rest, you can batch them up.

NewellClark committed:
Implement suggested changes

NewellClark committed:
Memorify RequestStream

- Memory overrides for System.Net.RequestStream
- Memory overrides for System.Net.NetworkStreamWrapper

NewellClark committed:
Spanified ChunkedMemoryStream

WriteAsync is implemented in terms of Write, so I went ahead and implemented Write(ReadOnlySpan<byte>). 
For some reason, AsSpan() isn't available in this file.

stephentoub commented on src/libraries/Common/src/System/IO/ChunkedMemoryStream.cs:
```diff
@@ -43,7 +50,9 @@ public override void Write(byte[] buffer, int offset, int count)
-                        Buffer.BlockCopy(buffer, offset, _currentChunk._buffer, _currentChunk._freeOffset, toCopy);
+                        ReadOnlySpan<byte> source = buffer.Slice(offset, toCopy);
+                        Span<byte> destination = new Span<byte>(_currentChunk._buffer, _currentChunk._freeOffset, toCopy);
+                        source.CopyTo(destination);
```
offset and count aren't needed any more, so this can be simplified a bit, e.g. something like:
```C#
while (!buffer.IsEmpty)
{
    if (_currentChunk is not null)
    {
        int remaining = _currentChunk._buffer.Length - _currentChunk._freeOffset;
        if (remaining > 0)
        {
            int toCopy = Math.Min(remaining, buffer.Length);

            buffer.Slice(0, toCopy).CopyTo(_currentChunk._buffer.AsSpan(_currentChunk._freeOffset));

            buffer = buffer.Slice(toCopy);
            _totalLength += toCopy;
            _currentChunk._freeOffset += toCopy;
            continue;
        }
    }

    AppendChunk(buffer.Length);
}
```

NewellClark replied:
I've made this change. I was unable to replace the span ctor with `AsSpan` due to the required assembly not being referenced by that project.

stephentoub replied:
Which project?

NewellClark replied:
System.Net.WebClient

stephentoub replied:
You can add System.Memory at https://github.com/dotnet/runtime/blob/665c65a8929a866b79409b832f0fa6af14cbf230/src/libraries/System.Net.WebClient/src/System.Net.WebClient.csproj#L32 if you'd like,  but using the ctor is also fine.

stephentoub commented on src/libraries/Common/src/System/IO/ChunkedMemoryStream.cs:
```diff
@@ -35,6 +35,13 @@ public byte[] ToArray()
 
         public override void Write(byte[] buffer, int offset, int count)
         {
+            Write(new ReadOnlySpan<byte>(buffer, offset, count));
```
This file doesn't have anything to do with the PR title of "DeflateStream.CopyToStream", does it?

NewellClark replied:
No, I need to change the PR title. I've put all the "trivial" stream changes in this PR.

stephentoub commented on src/libraries/System.IO.Compression/src/System/IO/Compression/DeflateZLib/Inflater.cs:
```diff
@@ -20,7 +21,7 @@ internal sealed class Inflater : IDisposable
         private readonly int _windowBits;                   // The WindowBits parameter passed to Inflater construction
         private ZLibNative.ZLibStreamHandle _zlibStream;    // The handle to the primary underlying zlib stream
-        private GCHandle _inputBufferHandle;                // The handle to the buffer that provides input to _zlibStream
+        private MemoryHandle _inputBufferHandle;       // The handle to the buffer that provides input to _zlibStream
```
```suggestion
        private MemoryHandle _inputBufferHandle;            // The handle to the buffer that provides input to _zlibStream
```

stephentoub commented on src/libraries/System.IO.Compression/src/System/IO/Compression/DeflateZLib/Inflater.cs:
```diff
@@ -180,16 +181,24 @@ public void SetInput(byte[] inputBuffer, int startIndex, int count)
+            Debug.Assert(!IsInputBufferHandleAllocated);
 
-            if (0 == count)
+            SetInput(inputBuffer.AsMemory().Slice(startIndex, count));
```
```suggestion
            SetInput(inputBuffer.AsMemory(startIndex, count));
```

stephentoub commented on src/libraries/System.Net.Requests/src/System/Net/RequestStream.cs:
```diff
@@ -105,6 +105,11 @@ public override Task WriteAsync(byte[] buffer, int offset, int count, Cancellati
+        {
+            return _buffer.WriteAsync(buffer, cancellationToken);
+        }
+
```
These System.Net.Requests changes aren't related to "DeflateStream.CopyToStream", are they?

NewellClark replied:
No, I updated PR title to clarify that.

NewellClark committed:
Apply suggested changes

NewellClark committed:
Merge remote-tracking branch 'dotnet/master' into DeflateStream.CopyToStream-memory

carlossanlop reviewed (approved):
Triage: Reviewed with @adamsitnik and @Jozkee  and it looks good.
Suggestions have been addressed and CI is passing.

adamsitnik commented on src/libraries/System.IO.Compression/src/System/IO/Compression/DeflateZLib/Inflater.cs:
```diff
@@ -313,14 +322,16 @@ private ZLibNative.ErrorCode Inflate(ZLibNative.FlushCode flushCode)
             }
         }
+
+        private unsafe bool IsInputBufferHandleAllocated => _inputBufferHandle.Pointer != (void*)0;
```
we should be able to use `default` here (and make the code look less like `C` lang ;) )

```suggestion
        private unsafe bool IsInputBufferHandleAllocated => _inputBufferHandle.Pointer != default;
```

adamsitnik commented on src/libraries/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs:
```diff
@@ -907,25 +907,38 @@ public void CopyFromSourceToDestination()
-                Debug.Assert(_deflateStream._inflater != null);
-                // Feed the data from base stream into the decompression engine.
-                _deflateStream._inflater.SetInput(buffer, offset, count);
+                return WriteAsyncCore(buffer.AsMemory().Slice(offset, count), cancellationToken).AsTask();
```
we can use the overload of `AsMemory` that accepts the offset and count and avoid one extra call to `Slice`

```suggestion
                return WriteAsyncCore(buffer.AsMemory(offset, count), cancellationToken).AsTask();
```

adamsitnik reviewed (approved):
LGTM, thank you @NewellClark !

adamsitnik committed:
Apply suggestions from code review
